gluon-web-admin: add option to show/hide password-login and add minimum password length

rubo77 opened the pull request:
Or maybe better min 12 characters?

It would be nice, if you get a verbose error message, if the pass is too short, but that needs some more work on gluon-web, i guess.

![pw](https://user-images.githubusercontent.com/1591563/27477720-477567c4-580d-11e7-8f9a-57b23d48cef9.png)

NeoRaider commented:
How about adding a hint about the minimum length to one the texts, for example the label of the text field?

Sunz3r commented:
pls let the user decide how strong the password should be

rubo77 commented:
If you want a weak password, you can set that ok the console.

If we allow weak passwords​ by default, we'll get complaints, that some users home net was hacked by Freifunk.

We could add a site. Conf option"allow weak passwords"

rubo77 commented:
We could add a site. Conf option"allow weak passwords"

rubo77 commented:
We could add a check for the common mistakes like repeating string, no capital chat, no number,....

Sunz3r commented:
you are right, hacked by freifunk are very bad.
how about limiting brute force attacks on ssh with iptables?

rubo77 commented:
I added to the text " (min. 8 characters)"

rubo77 commented:
tested and working. I added a screenshot to the first post.

the empty password field is now always red on load, which looks a bit strange.

NeoRaider commented:
I don't see a reason to allow passwords < 8 characters, that's a very weak limilation.

If you really want a weaker password, you can still set one via SSH.

rotanid commented:
good idea @rubo77 .
we removed the password option completely in our firmware - is just don't have time to upstream this using a site.conf-option to enable/disable this...

rotanid commented on package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua:
```diff
@@ -33,8 +33,9 @@ end
 	translate(
-                "Alternatively, you can set a password to access you node. Please choose a secure password you don't use anywhere else.<br /><br />"
-                .. "If you set an empty password, login via password will be disabled. This is the default."
+		"Alternatively, you can set a password to access you node. Please choose a "
```
although this typo was already there, please fix "you node" -> "your node"

rubo77 commented:
What would you want to control via site conf?

Maybe I find the time

rotanid commented:
@rubo77 control if the password-ssh-option is in the config mode or not.
we removed it in our gluon fork, but did it by removing the code, as that was easier than making it optional via site.conf
ask me on IRC if something isn't clear

rubo77 commented:
OK, i will add two options to site.conf:

remote = {
  allow_password_login = true, -- optional; default: true
  min_passwordlength = '8'     -- optional; default: 0; set this to 0 to disable a length check
},

rubo77 commented on package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua:
```diff
@@ -30,80 +31,88 @@ function keys:write(value)
 	end
 end
 
+if ((site.config_mode or {}).remote_login or {}).allow_password_login then
```
not sure, how to formulate this, so it defaults to true

rubo77 commented:
it is configurable now in site.conf.

rubo77 commented:
What do you thhink of an option to check the password-strength also?

I could add a codeblock that checks for existance of capital letters, numbers and special chars

rotanid commented on docs/package/gluon-web-admin.rst:
```diff
@@ -0,0 +1,27 @@
+  config mode is hidden
+  
+config_mode.remote_login.min_password_length \: optional (defaults to '8')
+  This stes a minimum limit to the allowed password length. Set this to 0 to
```
"stes" is wrong. also, the sentence is weird... maybe "This sets the minimum allowed password length."

rotanid commented:
checking the password strength should be optional though, if you add it.
suggestion would be to add some levels like "0 = no check", "1 = at least two of four types of chars" , "2 = at least three of four types" etc. - where types are lowercase letters, uppercase letters, digits, special chars

please also update the pull request title and text and the commit message according to your additional changes

NeoRaider commented:
Having an option to disable the password form completely make sense in my opinion; the other options just add complexity. We should define simple, sane defaults and keep it at that.

rubo77 commented:
But I want to set the minimum length to 12 characters in Kiel

NeoRaider commented:
Is this an actual issue? Passwords are only for people who need SSH access. If you know how to use the command line, you should be able to choose a good password...

If you're concerned about stupid users, you really shoud just hide the password form. It will still be possible to set passwords over SSH in config mode (and users who're annoyed by arbitrary password restrictions in the web interface will simply go that route...)

rubo77 commented:
OK, If I remove the length-definition, I will set the minimum Length to 12 characters

rubo77 commented:
OK, I removed the length-definition and set the minimum Length to 12 characters

kb-light commented on package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua:
```diff
@@ -30,80 +31,87 @@ function keys:write(value)
 	end
 end
 
+if ((site.config_mode or {}).remote_login or {}).allow_password_login ~= false then
```
maybe:
```
if ((site.config_mode or {}).remote_login or {}).allow_password_login == false then
 	-- password login is disabled in site.conf
 	return f_keys
end
[...]
return f_keys, f_password
```
to not indent the whole script

rubo77 replied:
I thought about this earlyer, but then I thought it wouldn't be good readeable to exit the script so early with return. But fine ;=

done

rubo77 commented:
I built it again, and it looks fine.

NeoRaider commented on package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua:
```diff
@@ -43,12 +48,15 @@ local s = f_password:section(Section)
 
-local pw2 = s:option(Value, "pw2", translate("Confirmation"))
+local pw2 = s:option(Value, "pw2", translate("Confirmation"),
+	translate("Minimum") .. " " .. 12 .. " " .. translate("characters"))
```
Always use `translatef` with format strings instead of splitting up sentences into multiple translation strings (sentences aren't constructed the same way in all languages). The English string should be "%u characters min.".

NeoRaider commented on docs/package/gluon-web-admin.rst:
```diff
@@ -0,0 +1,21 @@
+
+  config_mode = {
+    remote_login = {
+      allow_password_login = true
```
Maybe "show_password_form"? Password login is still allowed when the form is disabled.

NeoRaider commented on package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua:
```diff
@@ -43,12 +48,15 @@ local s = f_password:section(Section)
 
 local pw1 = s:option(Value, "pw1", translate("Password"))
 pw1.password = true
+pw1.datatype = 'minlength(12)'
```
After thinking about this a bit more, I'm okay with making the limit configurable in site.conf after all.

NeoRaider replied:
You'll also need to set *optional* on these fields to allow unsetting the password again. The *write* method must then be adjusted to check for *nil* values.

NeoRaider commented on package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua:
```diff
@@ -30,11 +31,15 @@ function keys:write(value)
 	end
 end
 
+if ((site.config_mode or {}).remote_login or {}).allow_password_login == false then
```
site.conf check for the new setting is missing

rubo77 replied:
I'll check that if I built again. what exactly do you have in mind here?

rubo77 commented:
I added everything and tested it again. build complete and it works fine:

![auswahl_100](https://user-images.githubusercontent.com/1591563/27702731-43db3d9e-5d05-11e7-933b-6c4a8141166b.png)

NeoRaider commented on package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua:
```diff
@@ -86,6 +94,11 @@ local function set_password(password)
 end
 
 function f_password:write()
+	if pw1.data:len() < min_password_length then
```
Using the datatype is much better than checking manually, as the fields will give direct feedback (background color) then. Nicer textual error messages are a TODO for future improvements in gluon-web.

We're only using the ugly errmessage because we can't express the "same value" constraint using datatypes, ideally, we'd change that to use a similar validation in the future, too.

rubo77 replied:
I had this before, but Then the password field was always red, until you typed something in.

The solution now is much more intuitive and uses the correct warning colors.

NeoRaider replied:
It was red because you didn't define the fields as optional.

rubo77 replied:
done

NeoRaider commented on package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua:
```diff
@@ -30,11 +31,17 @@ function keys:write(value)
 	end
 end
 
+local min_password_length = ((site.config_mode or {}).remote_login or {}).min_password_length or 12
```
I'd like to keep the default unchanged from the current situation, so make the default 1.

rubo77 replied:
But we try to fix an existing serious deficit here, so adding a minimum length is cruicial to security in Freifunk.

And there is no drawback for users, that used a short password before, because there cannot be no check of en existing password that is too short.

rubo77 replied:
Usually I would agree, but here we try to fix an existing serious security whole. So adding a minimum length is cruicial for the security of the nodes. I cannot think of a case, where a short password is needed.

Also here there is no drawback for users, that used a short password before, because there cannot be no check of en existing password that is too short.

NeoRaider commented on package/gluon-web-admin/i18n/de.po:
```diff
@@ -17,7 +17,7 @@ msgid "Advanced settings"
 
 msgid ""
-"Alternatively, you can set a password to access you node. Please choose a "
+"Alternatively, you can set a password to access your node. Please choose a "
```
This is an unrelated change that affects multiple files. Please split it out into a second commit.

rubo77 replied:
Sure, this is unrelated, but it only affects files, that we change in this PR anyway. Should I still split it off?

NeoRaider replied:
Yes.

rubo77 replied:
done

NeoRaider commented on package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua:
```diff
@@ -47,7 +54,8 @@ function pw1.cfgvalue()
 
-local pw2 = s:option(Value, "pw2", translate("Confirmation"))
+local pw2 = s:option(Value, "pw2", translate("Confirmation"),
+	translatef("%u characters min.", min_password_length))
```
The length hint should only be shown when min_password_length > 1.

rubo77 replied:
done

PowerPan commented:
I think every user should have the chance to select his own password.

Maybe it could be better to make a informative message, that the password with the length < $minlength is a weak password but if someone want to use it he should do this. 

It is not our responsibility to enforce security, but we should point it out

rubo77 commented:
As said before: If you want a weak password, you can set it on the console via SSH. And if you want to allow it in your Community in config mode, you can set the min-length to 1 in your `site.conf`.

We are discussing at the moment the default settings if nothing is set in `site.conf`. Those should be set on maximum security, which would be one of the following:

rubo77 commented:
We could set as default **password-login disabled**

rubo77 commented:
We could set a long password as default (10 or 12 chaaracters)

rotanid commented:
@rubo77 the last time we talked about this in IRC, as far as i remember we said "default disabled, default 12 chars", because we said existing setups aren't touched and everyone can change it in the site.conf if they really want

rubo77 commented:
Ok, I changed the default for `show_password_form` to false

rotanid commented:
maybe the example in the documentation should reflect the default values, not something random.
also, please update the commit message to reflect the password show/hide

rubo77 commented:
I changed the example, so it shows how to enable it with 12 chars. and changed the text afccordingly

commit message is now "gluon-web-admin: add option to show/hide password-login and add minimum password length"

NeoRaider commented on package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua:
```diff
@@ -30,11 +31,18 @@ function keys:write(value)
 	end
 end
 
+local min_password_length = ((site.config_mode or {}).remote_login or {}).min_password_length or 12
```
You could reduce the code duplication a bit by using a variable for the remote login config:
```lua
local config = (site.config_mode or {}).remote_login or {}
```

NeoRaider commented on package/gluon-web-admin/luasrc/lib/gluon/web/model/admin/remote.lua:
```diff
@@ -43,12 +51,20 @@ local s = f_password:section(Section)
 end
 
-local pw2 = s:option(Value, "pw2", translate("Confirmation"))
+local hint = nil;
```
Unneeded semicolon and initialization (a variable's value will default to nil when not specified)

NeoRaider commented:
I'm not sure what happened to the *hint* variable, and the coding style of the latest changes is bad, but I'll just fix it up myself...

rubo77 commented:
the variable hint was just used once, so I thought it would be better to directly add it in place. But ok, I'll keep in mind, that the readeablility is more important than saving a few characters in the code.

NeoRaider commented:
Merged with minor adjustments.

rotanid commented:
commit that was merged, for reference:
84d117ff73ac3a2099f851d0c4748cbfd1f41626
